Flaky test: TimeoutOnDeadServer

https://github.com/grpc/grpc-go/pull/4753/checks?check_run_id=3588365780

```
--- FAIL: Test (146.14s)
    --- FAIL: Test/TimeoutOnDeadServer (0.17s)
        end2end_test.go:947: Running test in tcp-clear-v1-balancer environment...
        tlogger.go:101: clientconn.go:175 [core] Channel Created  (t=+1.236807ms)
        tlogger.go:101: clientconn.go:252 [core] parsed scheme: "passthrough"  (t=+1.387108ms)
        tlogger.go:101: resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[{localhost:37303  <nil> 0 <nil>}] <nil> <nil>}  (t=+1.621509ms)
        tlogger.go:101: resolver_conn_wrapper.go:183 [core] Resolver state updated: {Addresses:[{Addr:localhost:37303 ServerName: Attributes:<nil> Type:0 Metadata:<nil>}] ServiceConfig:<nil> Attributes:<nil>} (resolver returned new addresses)  (t=+1.73871ms)
        tlogger.go:101: clientconn.go:728 [core] ClientConn switching balancer to "pick_first"  (t=+1.887411ms)
        tlogger.go:101: clientconn.go:748 [core] Channel switches to new LB policy "pick_first"  (t=+1.982812ms)
        tlogger.go:101: clientconn.go:789 [core] Subchannel Created  (t=+2.131813ms)
        tlogger.go:101: clientconn.go:789 [core] Subchannel(id:3366) created  (t=+2.168613ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to CONNECTING  (t=+2.399514ms)
        tlogger.go:101: clientconn.go:1253 [core] Subchannel picks a new address "localhost:37303" to connect  (t=+2.804417ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to CONNECTING  (t=+3.124419ms)
        tlogger.go:101: picker_wrapper.go:161 [core] blockingPicker: the picked transport is not ready, loop back to repick  (t=+3.34132ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to READY  (t=+5.289132ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to READY  (t=+5.647934ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to IDLE  (t=+9.501658ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to IDLE  (t=+9.84716ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to CONNECTING  (t=+10.477464ms)
        tlogger.go:101: clientconn.go:1253 [core] Subchannel picks a new address "localhost:37303" to connect  (t=+10.799666ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to CONNECTING  (t=+11.255668ms)
        tlogger.go:101: clientconn.go:1325 [core] grpc: addrConn.createTransport failed to connect to {localhost:37303 localhost:37303 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing dial tcp [::1]:37303: connect: connection refused"  (t=+13.513382ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to TRANSIENT_FAILURE  (t=+13.860684ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to TRANSIENT_FAILURE  (t=+14.157086ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+63.200286ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to SHUTDOWN  (t=+63.80319ms)
        tlogger.go:101: clientconn.go:1496 [core] Subchannel Deleted  (t=+64.106991ms)
        tlogger.go:101: clientconn.go:1496 [core] Subchanel(id:3366) deleted  (t=+64.206792ms)
        tlogger.go:101: clientconn.go:1098 [core] Channel Deleted  (t=+64.342093ms)
        tlogger.go:101: types.go:68 [channelz] attempt to delete child with id 3366 from a parent (id=3364) that doesn't currently exist  (t=+64.426993ms)
        tlogger.go:101: types.go:68 [channelz] attempt to delete child with id 3366 from a parent (id=3364) that doesn't currently exist  (t=+64.468894ms)
        end2end_test.go:947: Running test in tcp-tls-v1-balancer environment...
        tlogger.go:101: clientconn.go:175 [core] Channel Created  (t=+67.459712ms)
        tlogger.go:101: clientconn.go:252 [core] parsed scheme: "passthrough"  (t=+67.558812ms)
        tlogger.go:101: resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[{localhost:46479  <nil> 0 <nil>}] <nil> <nil>}  (t=+67.675813ms)
        tlogger.go:101: resolver_conn_wrapper.go:183 [core] Resolver state updated: {Addresses:[{Addr:localhost:46479 ServerName: Attributes:<nil> Type:0 Metadata:<nil>}] ServiceConfig:<nil> Attributes:<nil>} (resolver returned new addresses)  (t=+67.737214ms)
        tlogger.go:101: clientconn.go:728 [core] ClientConn switching balancer to "pick_first"  (t=+67.841114ms)
        tlogger.go:101: clientconn.go:748 [core] Channel switches to new LB policy "pick_first"  (t=+67.905415ms)
        tlogger.go:101: clientconn.go:789 [core] Subchannel Created  (t=+68.096116ms)
        tlogger.go:101: clientconn.go:789 [core] Subchannel(id:3372) created  (t=+68.147316ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to CONNECTING  (t=+68.364817ms)
        tlogger.go:101: picker_wrapper.go:161 [core] blockingPicker: the picked transport is not ready, loop back to repick  (t=+68.76302ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to CONNECTING  (t=+68.911521ms)
        tlogger.go:101: clientconn.go:1253 [core] Subchannel picks a new address "localhost:46479" to connect  (t=+69.133622ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to READY  (t=+99.86821ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to READY  (t=+100.352713ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to IDLE  (t=+104.669239ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to IDLE  (t=+105.006741ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to CONNECTING  (t=+105.558045ms)
        tlogger.go:101: clientconn.go:1253 [core] Subchannel picks a new address "localhost:46479" to connect  (t=+105.859447ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to CONNECTING  (t=+106.036148ms)
        tlogger.go:101: clientconn.go:1325 [core] grpc: addrConn.createTransport failed to connect to {localhost:46479 x.test.example.com <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing dial tcp [::1]:46479: connect: connection refused"  (t=+107.570957ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to TRANSIENT_FAILURE  (t=+107.857259ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to TRANSIENT_FAILURE  (t=+108.181361ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+157.882265ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to SHUTDOWN  (t=+158.426568ms)
        tlogger.go:101: clientconn.go:1496 [core] Subchannel Deleted  (t=+158.657469ms)
        tlogger.go:101: clientconn.go:1496 [core] Subchanel(id:3372) deleted  (t=+158.70087ms)
        tlogger.go:101: clientconn.go:1098 [core] Channel Deleted  (t=+158.864271ms)
        tlogger.go:101: types.go:68 [channelz] attempt to delete child with id 3372 from a parent (id=3371) that doesn't currently exist  (t=+158.988971ms)
        tlogger.go:101: types.go:68 [channelz] attempt to delete child with id 3372 from a parent (id=3371) that doesn't currently exist  (t=+159.096572ms)
        end2end_test.go:947: Running test in tcp-clear environment...
        tlogger.go:101: clientconn.go:175 [core] Channel Created  (t=+160.188079ms)
        tlogger.go:101: clientconn.go:252 [core] parsed scheme: "passthrough"  (t=+160.35548ms)
        tlogger.go:101: resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[{localhost:40369  <nil> 0 <nil>}] <nil> <nil>}  (t=+160.46068ms)
        tlogger.go:101: resolver_conn_wrapper.go:183 [core] Resolver state updated: {Addresses:[{Addr:localhost:40369 ServerName: Attributes:<nil> Type:0 Metadata:<nil>}] ServiceConfig:<nil> Attributes:<nil>} (resolver returned new addresses)  (t=+160.554081ms)
        tlogger.go:101: clientconn.go:789 [core] Subchannel Created  (t=+160.705982ms)
        tlogger.go:101: clientconn.go:789 [core] Subchannel(id:3377) created  (t=+160.768182ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to CONNECTING  (t=+161.117784ms)
        tlogger.go:101: clientconn.go:1253 [core] Subchannel picks a new address "localhost:40369" to connect  (t=+161.539587ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to CONNECTING  (t=+161.739188ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to READY  (t=+165.414611ms)
        tlogger.go:101: roundrobin.go:50 [roundrobin] roundrobinPicker: Build called with info: {map[0xc0006095e0:{{localhost:40369  <nil> 0 <nil>}}]}  (t=+165.789313ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to IDLE  (t=+171.712549ms)
        end2end_test.go:974: TestService/EmptyCall(context.Background.WithDeadline(2021-09-13 14:49:43.165555235 +0000 UTC m=+338.765299609 [-500.009µs]), _) = _, rpc error: code = Unavailable desc = error reading from server: EOF, want _, error code: DeadlineExceeded
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+172.590454ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to SHUTDOWN  (t=+173.074557ms)
        tlogger.go:101: clientconn.go:1496 [core] Subchannel Deleted  (t=+173.238758ms)
        tlogger.go:101: clientconn.go:1496 [core] Subchanel(id:3377) deleted  (t=+173.298559ms)
        tlogger.go:101: clientconn.go:1098 [core] Channel Deleted  (t=+173.398559ms)
        tlogger.go:101: types.go:68 [channelz] attempt to delete child with id 3377 from a parent (id=3376) that doesn't currently exist  (t=+173.46246ms)
        tlogger.go:101: types.go:68 [channelz] attempt to delete child with id 3377 from a parent (id=3376) that doesn't currently exist  (t=+173.53426ms)
FAIL
FAIL	google.golang.org/grpc/test	340.952s
```

--- a/test/channelz_test.go
+++ b/test/channelz_test.go
@@ -1531,21 +1531,11 @@ func (s) TestCZSubChannelTraceCreationDeletion(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Wait for ready
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
-	for src := te.cc.GetState(); src != connectivity.Ready; src = te.cc.GetState() {
-		if !te.cc.WaitForStateChange(ctx, src) {
-			t.Fatalf("timed out waiting for state change.  got %v; want %v", src, connectivity.Ready)
-		}
-	}
+	awaitState(ctx, t, te.cc, connectivity.Ready)
 	r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: "fake address"}}})
-	// Wait for not-ready.
-	for src := te.cc.GetState(); src == connectivity.Ready; src = te.cc.GetState() {
-		if !te.cc.WaitForStateChange(ctx, src) {
-			t.Fatalf("timed out waiting for state change.  got %v; want !%v", src, connectivity.Ready)
-		}
-	}
+	awaitNotState(ctx, t, te.cc, connectivity.Ready)
 
 	if err := verifyResultWithDelay(func() (bool, error) {
 		tcs, _ := channelz.GetTopChannels(0, 0)
@@ -2016,21 +2006,11 @@ func (s) TestCZTraceOverwriteSubChannelDeletion(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Wait for ready
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
-	for src := te.cc.GetState(); src != connectivity.Ready; src = te.cc.GetState() {
-		if !te.cc.WaitForStateChange(ctx, src) {
-			t.Fatalf("timed out waiting for state change.  got %v; want %v", src, connectivity.Ready)
-		}
-	}
+	awaitState(ctx, t, te.cc, connectivity.Ready)
 	r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: "fake address"}}})
-	// Wait for not-ready.
-	for src := te.cc.GetState(); src == connectivity.Ready; src = te.cc.GetState() {
-		if !te.cc.WaitForStateChange(ctx, src) {
-			t.Fatalf("timed out waiting for state change.  got %v; want !%v", src, connectivity.Ready)
-		}
-	}
+	awaitNotState(ctx, t, te.cc, connectivity.Ready)
 
 	// verify that the subchannel no longer exist due to trace referencing it got overwritten.
 	if err := verifyResultWithDelay(func() (bool, error) {

--- a/test/clientconn_state_transition_test.go
+++ b/test/clientconn_state_transition_test.go
@@ -519,3 +519,21 @@ func stayConnected(ctx context.Context, cc *grpc.ClientConn) {
 		}
 	}
 }
+
+func awaitState(ctx context.Context, t *testing.T, cc *grpc.ClientConn, stateWant connectivity.State) {
+	t.Helper()
+	for state := cc.GetState(); state != stateWant; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatalf("timed out waiting for state change.  got %v; want %v", state, stateWant)
+		}
+	}
+}
+
+func awaitNotState(ctx context.Context, t *testing.T, cc *grpc.ClientConn, stateDoNotWant connectivity.State) {
+	t.Helper()
+	for state := cc.GetState(); state == stateDoNotWant; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatalf("timed out waiting for state change.  got %v; want NOT %v", state, stateDoNotWant)
+		}
+	}
+}

--- a/test/creds_test.go
+++ b/test/creds_test.go
@@ -200,7 +200,7 @@ func (s) TestGRPCMethodAccessibleToCredsViaContextRequestInfo(t *testing.T) {
 	cc := te.clientConn(grpc.WithPerRPCCredentials(&methodTestCreds{}))
 	tc := testgrpc.NewTestServiceClient(cc)
 
-	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
 	if _, err := tc.EmptyCall(ctx, &testpb.Empty{}); status.Convert(err).Message() != wantMethod {
 		t.Fatalf("ss.client.EmptyCall(_, _) = _, %v; want _, _.Message()=%q", err, wantMethod)
@@ -233,7 +233,7 @@ func (s) TestFailFastRPCErrorOnBadCertificates(t *testing.T) {
 	defer te.tearDown()
 
 	opts := []grpc.DialOption{grpc.WithTransportCredentials(clientAlwaysFailCred{})}
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
 	cc, err := grpc.DialContext(ctx, te.srvAddr, opts...)
 	if err != nil {
@@ -261,17 +261,15 @@ func (s) TestWaitForReadyRPCErrorOnBadCertificates(t *testing.T) {
 	defer te.tearDown()
 
 	opts := []grpc.DialOption{grpc.WithTransportCredentials(clientAlwaysFailCred{})}
-	dctx, dcancel := context.WithTimeout(context.Background(), 10*time.Second)
-	defer dcancel()
-	cc, err := grpc.DialContext(dctx, te.srvAddr, opts...)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	cc, err := grpc.DialContext(ctx, te.srvAddr, opts...)
 	if err != nil {
 		t.Fatalf("Dial(_) = %v, want %v", err, nil)
 	}
 	defer cc.Close()
 
 	tc := testgrpc.NewTestServiceClient(cc)
-	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
-	defer cancel()
 	if _, err = tc.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); strings.Contains(err.Error(), clientAlwaysFailCredErrorMsg) {
 		return
 	}
@@ -444,17 +442,9 @@ func (s) TestCredsHandshakeAuthority(t *testing.T) {
 	defer cc.Close()
 	r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: lis.Addr().String()}}})
 
-	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	for {
-		s := cc.GetState()
-		if s == connectivity.Ready {
-			break
-		}
-		if !cc.WaitForStateChange(ctx, s) {
-			t.Fatalf("ClientConn is not ready after 100 ms")
-		}
-	}
+	awaitState(ctx, t, cc, connectivity.Ready)
 
 	if cred.got != testAuthority {
 		t.Fatalf("client creds got authority: %q, want: %q", cred.got, testAuthority)
@@ -484,17 +474,9 @@ func (s) TestCredsHandshakeServerNameAuthority(t *testing.T) {
 	defer cc.Close()
 	r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: lis.Addr().String(), ServerName: testServerName}}})
 
-	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	for {
-		s := cc.GetState()
-		if s == connectivity.Ready {
-			break
-		}
-		if !cc.WaitForStateChange(ctx, s) {
-			t.Fatalf("ClientConn is not ready after 100 ms")
-		}
-	}
+	awaitState(ctx, t, cc, connectivity.Ready)
 
 	if cred.got != testServerName {
 		t.Fatalf("client creds got authority: %q, want: %q", cred.got, testAuthority)

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -965,38 +965,25 @@ func (s) TestTimeoutOnDeadServer(t *testing.T) {
 func testTimeoutOnDeadServer(t *testing.T, e env) {
 	te := newTest(t, e)
 	te.userAgent = testAppUA
-	te.declareLogNoise(
-		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
-		"grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing",
-		"grpc: addrConn.resetTransport failed to create client transport: connection error",
-	)
 	te.startServer(&testServer{security: e.security})
 	defer te.tearDown()
 
 	cc := te.clientConn()
 	tc := testgrpc.NewTestServiceClient(cc)
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
 	if _, err := tc.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil {
 		t.Fatalf("TestService/EmptyCall(_, _) = _, %v, want _, <nil>", err)
 	}
+	// Wait for the client to report READY, stop the server, then wait for the
+	// client to notice the connection is gone.
+	awaitState(ctx, t, cc, connectivity.Ready)
 	te.srv.Stop()
-	cancel()
-
-	// Wait for the client to notice the connection is gone.
-	ctx, cancel = context.WithTimeout(context.Background(), 500*time.Millisecond)
-	state := cc.GetState()
-	for ; state == connectivity.Ready && cc.WaitForStateChange(ctx, state); state = cc.GetState() {
-	}
-	cancel()
-	if state == connectivity.Ready {
-		t.Fatalf("Timed out waiting for non-ready state")
-	}
-	ctx, cancel = context.WithTimeout(context.Background(), time.Millisecond)
+	awaitNotState(ctx, t, cc, connectivity.Ready)
+	ctx, cancel = context.WithTimeout(ctx, 5*time.Millisecond)
 	_, err := tc.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true))
 	cancel()
-	if e.balancer != "" && status.Code(err) != codes.DeadlineExceeded {
-		// If e.balancer == nil, the ac will stop reconnecting because the dialer returns non-temp error,
-		// the error will be an internal error.
+	if status.Code(err) != codes.DeadlineExceeded {
 		t.Fatalf("TestService/EmptyCall(%v, _) = _, %v, want _, error code: %s", ctx, err, codes.DeadlineExceeded)
 	}
 	awaitNewConnLogOutput()
@@ -1070,11 +1057,6 @@ func (s) TestFailFast(t *testing.T) {
 func testFailFast(t *testing.T, e env) {
 	te := newTest(t, e)
 	te.userAgent = testAppUA
-	te.declareLogNoise(
-		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
-		"grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing",
-		"grpc: addrConn.resetTransport failed to create client transport: connection error",
-	)
 	te.startServer(&testServer{security: e.security})
 	defer te.tearDown()
 
@@ -1114,9 +1096,6 @@ func testServiceConfigSetup(t *testing.T, e env) *test {
 	te := newTest(t, e)
 	te.userAgent = testAppUA
 	te.declareLogNoise(
-		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
-		"grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing",
-		"grpc: addrConn.resetTransport failed to create client transport: connection error",
 		"Failed to dial : context canceled; please retry.",
 	)
 	return te
@@ -1746,9 +1725,6 @@ func testPreloaderClientSend(t *testing.T, e env) {
 	te := newTest(t, e)
 	te.userAgent = testAppUA
 	te.declareLogNoise(
-		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
-		"grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing",
-		"grpc: addrConn.resetTransport failed to create client transport: connection error",
 		"Failed to dial : context canceled; please retry.",
 	)
 	te.startServer(&testServer{security: e.security})
@@ -1875,9 +1851,6 @@ func testMaxMsgSizeClientDefault(t *testing.T, e env) {
 	te := newTest(t, e)
 	te.userAgent = testAppUA
 	te.declareLogNoise(
-		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
-		"grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing",
-		"grpc: addrConn.resetTransport failed to create client transport: connection error",
 		"Failed to dial : context canceled; please retry.",
 	)
 	te.startServer(&testServer{security: e.security})
@@ -1942,9 +1915,6 @@ func testMaxMsgSizeClientAPI(t *testing.T, e env) {
 	te.maxClientReceiveMsgSize = newInt(1024)
 	te.maxClientSendMsgSize = newInt(1024)
 	te.declareLogNoise(
-		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
-		"grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing",
-		"grpc: addrConn.resetTransport failed to create client transport: connection error",
 		"Failed to dial : context canceled; please retry.",
 	)
 	te.startServer(&testServer{security: e.security})
@@ -2030,9 +2000,6 @@ func testMaxMsgSizeServerAPI(t *testing.T, e env) {
 	te.maxServerReceiveMsgSize = newInt(1024)
 	te.maxServerSendMsgSize = newInt(1024)
 	te.declareLogNoise(
-		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
-		"grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing",
-		"grpc: addrConn.resetTransport failed to create client transport: connection error",
 		"Failed to dial : context canceled; please retry.",
 	)
 	te.startServer(&testServer{security: e.security})
@@ -2141,11 +2108,6 @@ func testTap(t *testing.T, e env) {
 	te.userAgent = testAppUA
 	ttap := &myTap{}
 	te.tapHandle = ttap.handle
-	te.declareLogNoise(
-		"transport: http2Client.notifyError got notified that the client transport was broken EOF",
-		"grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing",
-		"grpc: addrConn.resetTransport failed to create client transport: connection error",
-	)
 	te.startServer(&testServer{security: e.security})
 	defer te.tearDown()
 
@@ -4878,17 +4840,9 @@ func testWaitForReadyConnection(t *testing.T, e env) {
 
 	cc := te.clientConn() // Non-blocking dial.
 	tc := testgrpc.NewTestServiceClient(cc)
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	state := cc.GetState()
-	// Wait for connection to be Ready.
-	for ; state != connectivity.Ready && cc.WaitForStateChange(ctx, state); state = cc.GetState() {
-	}
-	if state != connectivity.Ready {
-		t.Fatalf("Want connection state to be Ready, got %v", state)
-	}
-	ctx, cancel = context.WithTimeout(context.Background(), time.Second)
-	defer cancel()
+	awaitState(ctx, t, cc, connectivity.Ready)
 	// Make a fail-fast RPC.
 	if _, err := tc.EmptyCall(ctx, &testpb.Empty{}); err != nil {
 		t.Fatalf("TestService/EmptyCall(_,_) = _, %v, want _, nil", err)

--- a/test/goaway_test.go
+++ b/test/goaway_test.go
@@ -594,12 +594,7 @@ func (s) TestGoAwayThenClose(t *testing.T) {
 	client := testgrpc.NewTestServiceClient(cc)
 
 	t.Log("Waiting for the ClientConn to enter READY state.")
-	state := cc.GetState()
-	for ; state != connectivity.Ready && cc.WaitForStateChange(ctx, state); state = cc.GetState() {
-	}
-	if ctx.Err() != nil {
-		t.Fatalf("timed out waiting for READY channel state; last state = %v", state)
-	}
+	awaitState(ctx, t, cc, connectivity.Ready)
 
 	// We make a streaming RPC and do an one-message-round-trip to make sure
 	// it's created on connection 1.
@@ -622,11 +617,7 @@ func (s) TestGoAwayThenClose(t *testing.T) {
 	go s1.GracefulStop()
 
 	t.Log("Waiting for the ClientConn to enter IDLE state.")
-	for ; state != connectivity.Idle && cc.WaitForStateChange(ctx, state); state = cc.GetState() {
-	}
-	if ctx.Err() != nil {
-		t.Fatalf("timed out waiting for IDLE channel state; last state = %v", state)
-	}
+	awaitState(ctx, t, cc, connectivity.Idle)
 
 	t.Log("Performing another RPC to create a connection to server 2.")
 	if _, err := client.UnaryCall(ctx, &testpb.SimpleRequest{}); err != nil {

--- a/test/healthcheck_test.go
+++ b/test/healthcheck_test.go
@@ -212,44 +212,33 @@ func (s) TestHealthCheckWatchStateChange(t *testing.T) {
 
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	if ok := cc.WaitForStateChange(ctx, connectivity.Idle); !ok {
-		t.Fatal("ClientConn is still in IDLE state when the context times out.")
-	}
-	if ok := cc.WaitForStateChange(ctx, connectivity.Connecting); !ok {
-		t.Fatal("ClientConn is still in CONNECTING state when the context times out.")
-	}
+	awaitNotState(ctx, t, cc, connectivity.Idle)
+	awaitNotState(ctx, t, cc, connectivity.Connecting)
+	awaitState(ctx, t, cc, connectivity.TransientFailure)
 	if s := cc.GetState(); s != connectivity.TransientFailure {
 		t.Fatalf("ClientConn is in %v state, want TRANSIENT FAILURE", s)
 	}
 
 	ts.SetServingStatus("foo", healthpb.HealthCheckResponse_SERVING)
-	if ok := cc.WaitForStateChange(ctx, connectivity.TransientFailure); !ok {
-		t.Fatal("ClientConn is still in TRANSIENT FAILURE state when the context times out.")
-	}
+	awaitNotState(ctx, t, cc, connectivity.TransientFailure)
 	if s := cc.GetState(); s != connectivity.Ready {
 		t.Fatalf("ClientConn is in %v state, want READY", s)
 	}
 
 	ts.SetServingStatus("foo", healthpb.HealthCheckResponse_SERVICE_UNKNOWN)
-	if ok := cc.WaitForStateChange(ctx, connectivity.Ready); !ok {
-		t.Fatal("ClientConn is still in READY state when the context times out.")
-	}
+	awaitNotState(ctx, t, cc, connectivity.Ready)
 	if s := cc.GetState(); s != connectivity.TransientFailure {
 		t.Fatalf("ClientConn is in %v state, want TRANSIENT FAILURE", s)
 	}
 
 	ts.SetServingStatus("foo", healthpb.HealthCheckResponse_SERVING)
-	if ok := cc.WaitForStateChange(ctx, connectivity.TransientFailure); !ok {
-		t.Fatal("ClientConn is still in TRANSIENT FAILURE state when the context times out.")
-	}
+	awaitNotState(ctx, t, cc, connectivity.TransientFailure)
 	if s := cc.GetState(); s != connectivity.Ready {
 		t.Fatalf("ClientConn is in %v state, want READY", s)
 	}
 
 	ts.SetServingStatus("foo", healthpb.HealthCheckResponse_UNKNOWN)
-	if ok := cc.WaitForStateChange(ctx, connectivity.Ready); !ok {
-		t.Fatal("ClientConn is still in READY state when the context times out.")
-	}
+	awaitNotState(ctx, t, cc, connectivity.Ready)
 	if s := cc.GetState(); s != connectivity.TransientFailure {
 		t.Fatalf("ClientConn is in %v state, want TRANSIENT FAILURE", s)
 	}
@@ -278,12 +267,8 @@ func (s) TestHealthCheckHealthServerNotRegistered(t *testing.T) {
 
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	if ok := cc.WaitForStateChange(ctx, connectivity.Idle); !ok {
-		t.Fatal("ClientConn is still in IDLE state when the context times out.")
-	}
-	if ok := cc.WaitForStateChange(ctx, connectivity.Connecting); !ok {
-		t.Fatal("ClientConn is still in CONNECTING state when the context times out.")
-	}
+	awaitNotState(ctx, t, cc, connectivity.Idle)
+	awaitNotState(ctx, t, cc, connectivity.Connecting)
 	if s := cc.GetState(); s != connectivity.Ready {
 		t.Fatalf("ClientConn is in %v state, want READY", s)
 	}

--- a/test/pickfirst_test.go
+++ b/test/pickfirst_test.go
@@ -250,11 +250,7 @@ func (s) TestPickFirst_NewAddressWhileBlocking(t *testing.T) {
 	// Send a resolver update with no addresses. This should push the channel into
 	// TransientFailure.
 	r.UpdateState(resolver.State{})
-	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
-		if !cc.WaitForStateChange(ctx, state) {
-			t.Fatalf("timeout waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
-		}
-	}
+	awaitState(ctx, t, cc, connectivity.TransientFailure)
 
 	doneCh := make(chan struct{})
 	client := testgrpc.NewTestServiceClient(cc)

--- a/test/roundrobin_test.go
+++ b/test/roundrobin_test.go
@@ -119,11 +119,7 @@ func (s) TestRoundRobin_AddressesRemoved(t *testing.T) {
 	// Send a resolver update with no addresses. This should push the channel into
 	// TransientFailure.
 	r.UpdateState(resolver.State{Addresses: []resolver.Address{}})
-	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
-		if !cc.WaitForStateChange(ctx, state) {
-			t.Fatalf("timeout waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
-		}
-	}
+	awaitState(ctx, t, cc, connectivity.TransientFailure)
 
 	const msgWant = "produced zero addresses"
 	client := testgrpc.NewTestServiceClient(cc)
@@ -145,11 +141,7 @@ func (s) TestRoundRobin_NewAddressWhileBlocking(t *testing.T) {
 	// Send a resolver update with no addresses. This should push the channel into
 	// TransientFailure.
 	r.UpdateState(resolver.State{Addresses: []resolver.Address{}})
-	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
-		if !cc.WaitForStateChange(ctx, state) {
-			t.Fatalf("timeout waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
-		}
-	}
+	awaitState(ctx, t, cc, connectivity.TransientFailure)
 
 	client := testgrpc.NewTestServiceClient(cc)
 	doneCh := make(chan struct{})
@@ -229,12 +221,7 @@ func (s) TestRoundRobin_AllServersDown(t *testing.T) {
 		b.Stop()
 	}
 
-	// Wait for TransientFailure.
-	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
-		if !cc.WaitForStateChange(ctx, state) {
-			t.Fatalf("timeout waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
-		}
-	}
+	awaitState(ctx, t, cc, connectivity.TransientFailure)
 
 	// Failfast RPCs should fail with Unavailable.
 	client := testgrpc.NewTestServiceClient(cc)

